haystack/preview/telemetry/_telemetry.py

```diff
@@ -0,0 +1,158 @@
+            )
+            CONFIG_PATH.parents[0].mkdir(parents=True, exist_ok=True)
+            self.user_id = str(uuid.uuid4())
+            try:
+                with open(CONFIG_PATH, "w") as outfile:
+                    yaml.dump({"user_id": self.user_id}, outfile, default_flow_style=False)
+            except Exception as e:
+                logger.debug("Telemetry could not write config file to %s", CONFIG_PATH, exc_info=e)
+
+        self.event_properties = static_system_specs()
+
+    def send_event(self, event_name: str, event_properties: Optional[Dict[str, Any]] = None):
+        """
+        Sends a telemetry event.
+
+        :param event_name: The name of the event to show in PostHog.
+        :param event_properties: Additional event metadata. These are merged with the
+            system metadata collected in __init__, so take care not to overwrite them.
+        """
+        event_properties = event_properties or {}
+        dynamic_specs = dynamic_system_specs()
+        try:
+            posthog.capture(
+                distinct_id=self.user_id,
+                event=event_name,
+                properties={**self.event_properties, **dynamic_specs, **event_properties},
+            )
+        except Exception as e:
+            logger.debug("Telemetry couldn't make a POST request to PostHog.", exc_info=e)
+
+
+def send_telemetry(func):
+    """
+    Decorator that sends the output of the wrapped function to PostHog.
+    The wrapped function is actually called only if telemetry is enabled.
+    """
+
+    # FIXME? Somehow, functools.wraps makes `telemetry` out of scope. Let's take care of it later.
+    def send_telemetry_wrapper(*args, **kwargs):
+        try:
+            if telemetry:
+                output = func(*args, **kwargs)
+                if output:
+                    telemetry.send_event(*output)
+        except Exception as e:
+            # Never let telemetry break things
+            logger.debug("There was an issue sending a telemetry event", exc_info=e)
+
+    return send_telemetry_wrapper
+
+
+@send_telemetry
+def pipeline_running(pipeline: "Pipeline") -> Optional[Tuple[str, Dict[str, Any]]]:
+    """
+    Collects name, type and the content of the _telemetry_data attribute, if present, for each component in the
+    pipeline and sends such data to Posthog.
+
+    :param pipeline: the pipeline that is running.
+    """
+    pipeline._telemetry_runs += 1
+    if pipeline._telemetry_runs != 1 or pipeline._telemetry_runs % PIPELINE_RUN_BUFFER_SIZE != 0:
+        return None
+
+    # Collect info about components
+    pipeline_description = pipeline.to_dict()
+    components: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
+    for component_name, component in pipeline_description["components"].items():
+        components[component["type"]].append({"name": component_name})
+
+    # Data sent to Posthog
+    return "Pipeline run (2.x)", {
+        "pipeline_id": str(id(pipeline)),
```

Comment:
Ideally this would be the same pipeline_id if a pipeline with the exact same configuration is run later again, or even on a different machine. Can't we create a hash based on `pipeline_description`? If it's out of scope for this PR, let's  at least create an issue.

Reply:
Happy to create the new issue :blush: